### ISSUE TITLE
Default parameter value of Empty String is not recognized in processing of Intrinsic function

### ISSUE THREAD

**Issue by michaelj-smith** [awslabs/aws-sam-cli#1476]
### Description

First off, thank you so much for the support of Intrinsic Functions provided in #1261 ! This is really excellent work!

I've found a minor bug in the processing of Intrinsics when referencing a parameter with a default value of the empty string (""). Instead of interpolating the empty string, the intrinsic processesor interpolates the name of the parameter.

### Steps to reproduce

Sample SAM YML template in ./test.sam.yml:
NOTE: the "paramFeatureBranch" parameter has a default value of ""

```
AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Parameters:
  paramEnvironment:
    Type: String
    Description: Which environment do you want to deploy to? (local,dev,stage, or prod)
    AllowedValues:
    - local
    - dev
    - stage
    - prod
    Default: local
  paramFeatureBranch:
    Type: String
    Description: Provide the name of the feature branch if this in not a build from the master code branch.
    Default: ""

Globals:
  Function:
    Runtime: nodejs10.x
    MemorySize: 512
    Timeout: 30
    Environment:
      Variables:
        dbTableName: !Sub "${paramEnvironment}${paramFeatureBranch}_Products"

Resources:
  resLambdaTest:
    Type: AWS::Serverless::Function
    Properties:
      Handler: test.handler
      FunctionName: !Sub "${paramEnvironment}${paramFeatureBranch}_test"
      CodeUri: src
      Events:
        api:
          Type: Api
          Properties:
            Path: /test
            Method: GET

```

And this sample handler file in src/test.js:

```
exports.handler = async () => {
  const response = {
    dbTableName: process.env.dbTableName
  };

  return {
    statusCode: 200,
    body: JSON.stringify(response)
  };
};
```

Execute the local api via this command:

```
sam local start-api --template test.sam.yml
```

Execute a curl/browser/Postman request to http://127.0.0.1:3000/test

### Observed result

API response is:

```
{
    "dbTableName": "localparamFeatureBranch_Products"
}
```

Where ${paramFeatureBranch} was interpolated as "paramFeatureBranch"

### Expected result

```
{
    "dbTableName": "local_Products"
}
```

where ${paramFeatureBranch} should have been interpolated as ""

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: macOS v10.14.6 (Mojave)
2. `sam --version`: SAM CLI, version 0.22.0

I also tried executing the local api using parameter-overrides with this command:

```
sam local start-api --template test.sam.yml --parameter-overrides "ParameterKey=paramFeatureBranch,ParameterValue=" 
```

but that fails with this error message:

> Error: Invalid value for "--parameter-overrides": ParameterKey=paramFeatureBranch,ParameterValue=  is not in valid format. It must look something like 'ParameterKey=KeyPairName,ParameterValue=MyKey ParameterKey=InstanceType,ParameterValue=t1.micro'

This is fine, I understand that passing in an empty string for parameter overrides may not be supported. However, I expect that a default value of empty string should be recognized.

Thanks for your support!

**Comment by michaelj-smith** [comment]
I'm pretty sure this bug can be corrected in samcli/commands/local/lib/sam_base_provider.py, line 156:

```
if isinstance(value, dict) and "Default" in value:
```

If the "Default" property has the value of "", then it will evaluate to false.

**Comment by jfuss** [comment]
@michaelj-smith Thanks for the detailed report. This does look like a bug at first glance. 

Can you try:
`sam local start-api --template test.sam.yml --parameter-overrides "ParameterKey=paramFeatureBranch,ParameterValue=''"` Not sure if that will force the empty string but worth a try.

> `if isinstance(value, dict) and "Default" in value:`

This checks to make sure the "Default" key is in the Dictionary before accessing its value.

The issue is within samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py, line 201. We currently are checking for `if translated:` which evaluates to false with an empty string. Instead this should be an explicit None check `if translated is not None`.

Patch: 
```
diff --git a/samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py b/samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py
index 036deef..4ee8082 100644
--- a/samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py
+++ b/samcli/lib/intrinsic_resolver/intrinsics_symbol_table.py
@@ -198,7 +198,7 @@ class IntrinsicsSymbolTable(object):
         """
         # pylint: disable-msg=too-many-return-statements
         translated = self.get_translation(logical_id, resource_attribute)
-        if translated:
+        if translated is not None:
             return translated

         if logical_id in self.SUPPORTED_PSEUDO_TYPES:
```

This will need an integ test to make sure we don't regress (one for `sam local invoke` is sufficient) and updated unit tests. @michaelj-smith Would you be willing to send a PR with those?

**Comment by michaelj-smith** [comment]
I'll work on it early next week. Thanks for the additional context!

**Comment by michaelj-smith** [comment]
Executing the suggested command:
```
sam local start-api --template test.sam.yml --parameter-overrides "ParameterKey=paramFeatureBranch,ParameterValue=''"
```

Produces this result:
```
{
"dbTableName": "local''_Products"
}
```

I also attempted to escape the double-quotes:
```
sam local start-api --template test.sam.yml --parameter-overrides "ParameterKey=paramFeatureBranch,ParameterValue=\"\""
```

which produces the original result:
```
{
"dbTableName": "localparamFeatureBranch_Products"
}
```

It appears that there isn't a suitable workaround with the parameter-overrides. I'll continue to work on a PR this week.

**Comment by jfuss** [comment]
@michaelj-smith :( Thanks for the update. Let me know if you need any help in the process.

**Comment by blutter** [comment]
@michaelj-smith I'm wondering if you have made any progress on your PR as I'm seeing a similar issue with an empty ParameterValue when deploying a stack.

**Comment by malimohub** [comment]
any progress on this issue?

**Comment by jfuss** [comment]
Was released in [v0.42.0](https://github.com/awslabs/aws-sam-cli/releases/tag/v0.42.0).

Closing
